packages/block-editor/src/hooks/block-binding-support/with-block-binding-support.js

```diff
@@ -0,0 +1,178 @@
+			} );
+		},
+		[ attrName, setAttributes ]
+	);
+
+	// Store a reference to the last value and attribute value.
+	const lastPropValue = useRef( propValue );
+	const lastAttrValue = useRef( attrValue );
+
+	/*
+	 * Initially sync (first render / onMount ) attribute
+	 * value with the source prop value.
+	 */
+	useEffect( () => {
+		updateBoundAttibute( propValue );
+	}, [ propValue, updateBoundAttibute ] );
+
+	/*
+	 * Sync data.
+	 * This effect will run every time
+	 * the attribute value or the prop value changes.
+	 * It will sync them in both directions.
+	 */
+	useEffect( () => {
+		/*
+		 * Source Prop => Block Attribute
+		 *
+		 * Detect changes in source prop value,
+		 * and update the attribute value accordingly.
+		 */
+		if ( typeof propValue !== 'undefined' ) {
+			if ( propValue !== lastPropValue.current ) {
+				lastPropValue.current = propValue;
+				updateBoundAttibute( propValue );
+				return;
+			}
+		} else if ( placeholder ) {
+			/*
+			 * If the attribute is `src` or `href`,
+			 * a placeholder can't be used because it is not a valid url.
+			 * Adding this workaround until
+			 * attributes and metadata fields types are improved and include `url`.
+			 */
+			const htmlAttribute =
+				getBlockType( blockName ).attributes[ attrName ].attribute;
+
+			if ( htmlAttribute === 'src' || htmlAttribute === 'href' ) {
+				updateBoundAttibute( null );
+				return;
+			}
+
+			updateBoundAttibute( placeholder );
+		}
+
+		/*
+		 * Block Attribute => Source Prop
+		 *
+		 * Detect changes in block attribute value,
+		 * and update the source prop value accordingly.
+		 */
+		if ( attrValue !== lastAttrValue.current && updatePropValue ) {
+			lastAttrValue.current = attrValue;
+			updatePropValue( attrValue );
+		}
+	}, [
+		updateBoundAttibute,
+		propValue,
+		attrValue,
+		updatePropValue,
+		placeholder,
+		blockName,
+		attrName,
+	] );
+
+	return null;
+};
+
+const withBlockBindingSupport = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const { attributes, name } = props;
+
+		const { getBlockBindingsSource } = unlock(
+			useSelect( blockEditorStore )
+		);
+
+		// Bail early if there are no bindings.
+		const bindings = attributes?.metadata?.bindings;
+		if ( ! bindings ) {
+			return <BlockEdit { ...props } />;
+		}
+
+		const BindingConnectorInstances = [];
+
+		Object.entries( bindings ).forEach( ( [ attrName, settings ], i ) => {
+			const source = getBlockBindingsSource( settings.source );
+
+			if ( source ) {
+				const { useSource } = source;
+				const attrValue = attributes[ attrName ];
+
+				// Create a unique key for the connector instance
+				const key = `${ settings.source }-${ name }-${ attrName }-${ i }`;
+
+				BindingConnectorInstances.push(
+					<BlockBindingConnector
+						key={ key }
+						attrName={ attrName }
+						attrValue={ attrValue }
+						useSource={ useSource }
+						blockProps={ props }
+						args={ settings.args }
+					/>
+				);
+			}
+		} );
```

Comment:
Let's extract this in a separate component which only rendered when block has `bindings`.